app/src/main/java/com/pavesid/androidacademy/repositories/MoviesRepositoryImpl.kt

```diff
@@ -16,38 +18,64 @@ class MoviesRepositoryImpl @Inject constructor(
     private val moviesDao: MoviesDao
 ) : MoviesRepository {
 
-    override suspend fun getMovies(): List<Movie> = getMoviesFromDB().let {
-        if (it.isEmpty()) {
-            val list = getMoviesFromInternet()
-            insertMovies(list)
-            return@let list
-        } else it
-    }
+    private var genres = listOf<Genre>()
+
+    override suspend fun getMovies(
+        page: Int,
+        language: String
+    ): List<Movie> = getMoviesFromInternet(page, language)
+
+    override suspend fun getDetails(id: Int): Details? = getDetailsFromInternet(id)
+
+    override suspend fun getActors(id: Int): ActorsResponse = MoviesApiImpl.getActors(id)
 
     private suspend fun getMoviesFromDB(): List<Movie> = moviesDao.getAllMovies()
 
-    private suspend fun getMoviesFromInternet(): List<Movie> {
+    private suspend fun getMoviesFromInternet(
+        page: Int,
+        language: String
+    ): List<Movie> {
         var movies = listOf<JsonMovie>()
-        var genres = listOf<Genre>()
-        var actors = listOf<Actor>()
         coroutineScope {
             launch {
-                movies = MoviesApiImpl.getMovies()
+                movies = MoviesApiImpl.getMovies(page, language).movies
             }
-            launch {
-                genres = MoviesApiImpl.getGenres()
-            }
-            launch {
-                actors = MoviesApiImpl.getActors()
+            if (genres.isEmpty()) {
+                launch {
+                    genres = MoviesApiImpl.getGenres().genres
+                }
             }
         }
         return parseMovies(
             movies,
-            genres,
-            actors
+            genres
         )
     }
 
+    private suspend fun getDetailsFromInternet(
+        id: Int
+    ): Details? {
+        var details: DetailsResponse? = null
+        var actors: ActorsResponse? = null
+        coroutineScope {
+            launch {
+                details = MoviesApiImpl.getDetails(id)
+            }
+            launch {
+                actors = MoviesApiImpl.getActors(id)
+            }
+        }
+        return if (details != null && actors != null) {
+            Details(
+                details!!,
+                actors!!.cast,
+                actors!!.crew
+            )
```

Comment:
you can use "lateinit var". because, if you throw an exception inside any of the launch - your exceptionHandler will catch it and the parent coroutine job will be canceled.